Issue #1811: Add static methods to construct Future class using Calla…

v1ctor opened the pull request:
…ble, Supplier, Runnable
#1811

danieldietrich commented on javaslang/src/main/java/javaslang/concurrent/Future.java:
```diff
@@ -323,6 +324,79 @@
+     * @return A new Future instance
+     * @throws NullPointerException if computation is null.
+     */
+    static <T> Future<T> ofSupplier(Supplier<? extends T> computation) {
```
Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

danieldietrich commented on javaslang/src/main/java/javaslang/concurrent/Future.java:
```diff
@@ -323,6 +324,79 @@
+     * @return A new Future instance
+     * @throws NullPointerException if one of executorService or computation is null.
+     */
+    static <T> Future<T> ofSupplier(ExecutorService executorService, Supplier<? extends T> computation) {
```
Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

danieldietrich commented on javaslang/src/main/java/javaslang/concurrent/Future.java:
```diff
@@ -323,6 +324,79 @@
+     * @return A new Future instance
+     * @throws NullPointerException if computation is null.
+     */
+    static <T> Future<T> ofCallable(Callable<? extends T> computation) {
```
Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

danieldietrich commented on javaslang/src/main/java/javaslang/concurrent/Future.java:
```diff
@@ -323,6 +324,79 @@
+     * @return A new Future instance
+     * @throws NullPointerException if one of executorService or computation is null.
+     */
+    static <T> Future<T> ofCallable(ExecutorService executorService, Callable<? extends T> computation) {
```
Please add a

```java
Objects.requireNonNull(computation, "computation is null");
```

v1ctor commented:
Hi Daniel, thanks for the feedback! First I added null checks to the code, but then I realized that as soon as these methods are wrappers around existing functions these checks will be executed inside. So I removed them. Do you want me to add them explicitly?

danieldietrich commented:
Hi

> Do you want me to add them explicitly?

yes, please. Only for the mentioned method references. The `paramName::methodName` method references throw before passed as new parameter to an existing function.

The executorService does not need to be checked again.

---

**Example:**

```java
import java.util.Objects;
import java.util.function.Supplier;

public class Test {

    public static void main(String[] args) {
        testSupplier(null);
    }

    static <T> void testSupplier(Supplier<T> s) {
        testCheckedSupplier(s::get); // <------ throws immediately if s is null
    }

    static <T> void testCheckedSupplier(Supplier<T> s) {
        Objects.requireNonNull(s, "s is null");
        System.out.println("ok");
    }

}
```

**Output:**

```java
Exception in thread "main" java.lang.NullPointerException
	at javaslang.Test.testSupplier(Test.java:18)
	at javaslang.Test.main(Test.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

v1ctor commented:
@danieldietrich Yes, you're totally right. I've missed this case. Thanks!

danieldietrich commented:
Thank you!

v1ctor commented:
Thank you :)

danieldietrich commented:
Fixes #1811
